packages/manager-react-components/src/hooks/datagrid/useIcebergV2.tsx

```diff
@@ -72,3 +84,53 @@ export function useResourcesIcebergV2<T = unknown>({
     status,
   };
 }
+
+type UseIcebergV2InfiniteQueryParams<T> = {
+  iceberg: Omit<IcebergFetchParamsV2, 'cursor'>;
+  query: Omit<
+    UseInfiniteQueryOptions,
+    'queryFn' | 'getNextPageParam' | 'initialPageParam'
+  > & {
+    select?: (data: InfiniteData<T, unknown>) => T[];
+  };
+  shouldFetchAll?: boolean;
+};
+
+export function useIcebergV2InfiniteQuery<T>({
+  iceberg: { pageSize, search, ...icebergOptions },
+  query: { queryKey, ...queryOptions },
+  shouldFetchAll = false,
+}: UseIcebergV2InfiniteQueryParams<T>) {
+  const query = useInfiniteQuery({
+    // return only flattened data without cursor by default to avoid
+    // unwanted rerenders on refetch if data is the same
+    select: (data) =>
+      data?.pages?.flatMap((page: UseInfiniteQueryResult<T[]>) => page?.data),
+    ...queryOptions,
+    initialPageParam: null,
+    queryKey: [
+      ...queryKey,
+      shouldFetchAll ? 'all' : '',
+      search ? `${search.key}=${search.value}` : '',
+    ].filter(Boolean),
+    queryFn: ({ pageParam }) =>
+      fetchIcebergV2<T[]>({
+        pageSize:
+          pageSize || (shouldFetchAll ? API_V2_MAX_PAGE_SIZE : defaultPageSize),
+        cursor: pageParam as string,
+        search,
+        ...icebergOptions,
+      }),
+    getNextPageParam: (lastPage) =>
+      (lastPage as IcebergFetchResultV2<T[]>).cursorNext,
+  });
+
+  useEffect(() => {
+    // fetch next page if needed
+    if (shouldFetchAll && query.hasNextPage && !query.isFetchingNextPage) {
+      query.fetchNextPage();
+    }
+  }, [query.data]);
```

Comment:
On first call you already know the size of your request so you could request them all instead of fetching page by page

Reply:
for APIV2 we don't know the size